Add method to return the set of course ids for a given array of lesson ids

fjorgemota opened the pull request:
### Changes proposed in this Pull Request

* Adds the method  `get_course_ids` to the class `Sensei_Lesson` to allow us to get, for a given array of lesson IDs, all course IDs for that array of lesson IDs;

### Testing instructions

1. Change to this branch and run unit tests;
2. Also, use the Testing Instructions on the PR 1357-gh-Automattic/sensei-pro to check if this PR is working as desired;

### Related

Related issue: 1080-gh-Automattic/sensei-pro

fjorgemota committed:
Add the method get_course_ids to get the course ids from a list of lesson IDs

fjorgemota committed:
Add unit test for Sensei Lesson's get_course_id

fjorgemota committed:
Use method instead of function to refer to methods on Sensei Lesson class

fjorgemota committed:
Remove unnecessary index assignment on foreach loop in test

fjorgemota committed:
Add test unit to verify get_course_ids functionality

fjorgemota committed:
Fix formatting of the test-class-lesson.php file

fjorgemota committed:
Add caching for the get_course_ids method

fjorgemota committed:
Disable PHPCS rules related to the direct database queries

fjorgemota committed:
Use sort instead of asort on get_course_ids

fjorgemota committed:
Add method to verify if get_course_ids cache works correctly

fjorgemota committed:
Fix test-class-lesson.php formatting

fjorgemota commented on includes/class-sensei-lesson.php:
```diff
@@ -3822,6 +3822,53 @@ public function get_course_id( $lesson_id ) {
+
+		sort( $lesson_ids, SORT_NUMERIC );
+
+		$cache_key     = 'lesson/get-course-ids/' . implode( ',', $lesson_ids );
```
Right now I'm not hashing on the lesson IDs part, but it's definitely something that can grow a little bit (as it can include all the lessons found for a given course, so thousands of lesson IDs)

Do you think I should apply a `sha1` or something here? I appreciate suggestions here.

renatho replied:
Good point! IMO, it would seem interesting to avoid a super big key. Maybe it could become an issue in some env? In some envs, `wp_cache` gets saved in the DB, and I'm not sure the maximum field size we'd have for this. But hashing I think would make it safer.

aaronfc replied:
Good catch Fernando, definitively I would do something to limit the cache key as this could grow too big. Hashing sounds good to me (sha1 or even md5).

alexsanford replied:
I agree that a hash would be nice. Since it doesn't have to be cryptographically secure, I think we can optimize for speed. So md5 seems like a good option to me.

fjorgemota replied:
Hey people. Thanks for your feedback! Changed in [bb7a10c](https://github.com/Automattic/sensei/pull/5294/commits/bb7a10cb5caa534f38cf643d09c17ff998361d2e).

fjorgemota commented on includes/class-sensei-lesson.php:
```diff
@@ -3822,6 +3822,53 @@ public function get_course_id( $lesson_id ) {
+
+		$cache_key     = 'lesson/get-course-ids/' . implode( ',', $lesson_ids );
+		$cache_group   = 'sensei';
+		$cached_result = wp_cache_get( $cache_key, $cache_group );
```
I'm using `wp_cache_get` and `wp_cache_set` as suggested by PHPCS thanks to our calls to `$wpdb`.

fjorgemota commented on includes/class-sensei-lesson.php:
```diff
@@ -3822,6 +3822,53 @@ public function get_course_id( $lesson_id ) {
+		$courses_by_lesson = array_fill_keys( $lesson_ids, false );
+		$placeholders      = implode( ', ', array_fill( 0, count( $lesson_ids ), '%d' ) );
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare
```
I dlsabled these checks here because the `$placeholders` are dynamically computed thanks to the variable number of `$lesson_ids`, and I didn't find a way to avoid the warnings related to it. I appreciate any suggestions here!

fjorgemota commented on includes/class-sensei-lesson.php:
```diff
@@ -3822,6 +3822,53 @@ public function get_course_id( $lesson_id ) {
+				$courses_by_lesson[ $result->lesson_id ] = $result->course_id;
+			}
+		}
+		wp_cache_set( $cache_key, $courses_by_lesson, $cache_group );
```
By default, I'm always caching `$courses_by_lesson`, and if, for instance, `$results` is related to an error. This might not be the optimal approach, tho, so I'm accepting suggestions here too.

renatho replied:
Good point! I think if there's an error, we shouldn't cache it, and maybe return an empty array?

alexsanford replied:
I agree that we probably shouldn't cache if there is an error.

fjorgemota replied:
Fixed in [3712ceb](https://github.com/Automattic/sensei/pull/5294/commits/3712cebaeded12651c453ca3b1d5df27bef93ff8). Thanks you all!

fjorgemota reviewed (commented):
Just some comments for documentation purposes. =)

renatho commented on includes/class-sensei-lesson.php:
```diff
@@ -3822,6 +3822,53 @@ public function get_course_id( $lesson_id ) {
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare
+		$query = $wpdb->prepare(
+			"
+				SELECT lesson.ID AS lesson_id, course.ID AS course_id FROM {$wpdb->posts} lesson
```
Do you think we could avoid a query with multiple tables here somehow? Just in case we use it in the future in some place like Learnomattic, where the tables can live in different places.

aaronfc replied:
Agree on that. I think something like this should work? Let me know if I am missing something.

```
SELECT post_id as lesson_id, meta_value as course_id FROM {$wpdb->postmeta} WHERE `meta_key` = '_lesson_course' AND `post_id` IN ({$placeholders}) 
```

Where you would need to check for empty course_id later in case you want to set them with false or something special for lessons without course.

alexsanford replied:
I think part of the issue here is validation. If we can `JOIN` then we can check that the Course ID is valid, without the need to loop through them all and look for that course in the DB (which is part of the reason for the original perf issue).

> Just in case we use it in the future in some place like Learnomattic, where the tables can live in different places.

I remember hearing about this...but it seems pretty strange to me that the DB would be set up in such a way that `JOIN` queries are impossible. Are we sure it's possible for the `posts` and `postmeta` table to be in different locations for a single site?

I could understand, for example, if tables for different sites were in different locations, or if shared tables, like `users`, were stored in a different location. But it seems like `JOIN`'s should be possible between `posts` and `postmeta` 🤷‍♂️

renatho replied:
Hum! I think @alexsanford is right! Probably it's not the case for `posts` vs `postmeta`. But I can check it later to confirm since I'll work there today.

p1651851724334379-slack-C07418EJ0

fjorgemota replied:
Yep, as @alexsanford explained, the reasoning for this join is to check if the course iD is valid and refers to a course post type, without having to check again in the code. You can verify those checks on the `get_course_id` method, where that's checked individually.

I'm thinking here about the possibilities related to this, @renatho, as I know that you fixed a bug related to a similar JOIN on Learnomattic in the past, however, if I change this to avoid JOINs, we will basically do the same as a `get_course_id` in a loop (because we need to check if both course ID and lesson ID is valid), so...yeah.

alexsanford replied:
@renatho I just checked your slack thread, and as I suspected, the issue is with the `wp_users` table. That one is "special" on a multisite installation like dotcom, because all users across all sites are stored in the same table. It's probably quite large, and I can understand why it might need to be handled differently.

I imagine the reason that y'all couldn't find reference to the `wp_*_comments` table (I also couldn't find `wp_*_posts` or `wp_*_postmeta` for that matter) is because these tables are created (with a unique prefix) for each new site.

So I still think it's safe to assume that those tables should be safe to `JOIN`, even on dotcom, but would be nice to test/verify that 😅

fjorgemota replied:
I decided to ping Igor to check that. Here's the thread: p1656959044763259-slack-C07418EJ0

renatho replied:
Hey! Sorry for the delay in answering that, but I just checked it, and it works there. Sorry for the noise here!

aaronfc replied:
>I think part of the issue here is validation. If we can JOIN then we can check that the Course ID is valid, without the need to loop through them all and look for that course in the DB (which is part of the reason for the original perf issue).

@alexsanford I agree, but... what is it really validating? I mean, I can't imagine any "normal" flow that could make these values be something "different" from a course_id. And anyway, if it happened it could be that, by coincidence the "broken" id matched a "valid" course still... So... I am not sure these validations are adding much.
Maybe we want to ensure that courses are not "deleted" but still we are not validating that, as it could be that the status for the courses was "deleted".

Anyway I always prefered to avoid joins because some times are not really necessary and just overcomplicate queries. But I am ok with continuing with it if you all agree on it's need.

alexsanford replied:
> but... what is it really validating?

Unexpected weirdness that could break stuff 😅 

It may be unlikely for the validation to fail, but personally I think it's still worth doing. Never underestimate the ability for a WP database to get messed up haha.

> Maybe we want to ensure that courses are not "deleted" but still we are not validating that

That's a good point. CC @fjorgemota do you think we should add this status check? I'm not sure this is checked in `get_course_id` either, so maybe it's ok to leave this out?

fjorgemota replied:
Hey there.

Yeah, I don't think it's worth checking if a course is deleted here, especially because having a course associated is all we care for.

About the joins and the extra validation, yeah, I agree that wouldn't be needed in a normal environment, however, given the freedom we have on WordPress with post-meta, and no current way to set up a proper foreign key between the post types, I think the validation is...worth it. Better safe than sorry, as usual.

renatho commented on tests/unit-tests/test-class-lesson.php:
```diff
@@ -101,13 +100,110 @@ public function testIsPreRequisiteComplete() {
+		$lesson_ids             = $this->factory->lesson->create_many( 9 );
+		$lesson_id_to_course_id = array();
+		foreach ( $lesson_ids as $lesson_id ) {
+			$course_index                         = array_rand( $course_ids );
```
Would it make the test simpler and more predictable if instead of `array_rand` we used something like `$lesson_ids_index % 3`?

fjorgemota replied:
hmmm, good point. I changed it in [29ee76f](https://github.com/Automattic/sensei/pull/5294/commits/29ee76fe5f832d42930fff4fa99992fb3b473a19).

renatho replied:
My point of simplifying that is that we could remove the `$lesson_id_to_course_id` because since we'll know the respective course, we wouldn't need that, right?

And I think we just forgot the `testGetCourseId` that still has the  `array_rand`.

renatho replied:
Reading a little more the `testGetCourseId`, actually I don't think we need a loop here, right?
Maybe we could just create one course with a lesson, and check the `get_course_id`?

fjorgemota replied:
Hey there.

> My point of simplifying that is that we could remove the $lesson_id_to_course_id because since we'll know the respective > course, we wouldn't need that, right?

Fair enough. I adapted the code to avoid the usage of this variable and instead compute the expected course ID manually. Please note that this isn't as simple as a `%` operation, because this only returns the expected course index, so we need to get the course ID from the course list anyway. Here are some commits where I basically fix that: [f96947a](https://github.com/Automattic/sensei/pull/5294/commits/f96947a87d5afc503645565b242d1aae7a05d2a5) and [7b2a7c9](https://github.com/Automattic/sensei/pull/5294/commits/7b2a7c930e3c8b43fcbec3a2ce155166a31d5156) (I had to copy the lesson IDs array to avoid issues with the shuffled list, by the way).

> And I think we just forgot the testGetCourseId that still has the array_rand.

I fixed that in [84b345f](https://github.com/Automattic/sensei/pull/5294/commits/84b345f59fcc253f190ddbcc0bb933794088263c). The commit message refers only to the formatting of the PHP file, but it includes that change too.

> Reading a little more the `testGetCourseId`, actually I don't think we need a loop here, right?
> Maybe we could just create one course with a lesson, and check the `get_course_id`?

Actually, kinda. While for content drip specifically usually we always get the same course ID for every lesson queried, this method is meant to be...more generic, so my line of thinking is more like "if it can get a bunch of different courses IDs for different lessons IDs, then it will be able to determine the same course ID for a bunch of lesson IDs".

Does that make sense? Given that the API allows that, testing for that case looks like a better idea, IMHO.

renatho replied:
> Actually, kinda. While for content drip specifically usually we always get the same course ID for every lesson queried, this method is meant to be...more generic, so my line of thinking is more like "if it can get a bunch of different courses IDs for different lessons IDs, then it will be able to determine the same course ID for a bunch of lesson IDs".
> 
> Does that make sense? Given that the API allows that, testing for that case looks like a better idea, IMHO.

I don't agree because we're just testing a method that gets a post meta, and make some post type checks. But I can agree to disagree here. 😄

fjorgemota replied:
aaaahhhh, actually, thinking again, you are right, @renatho. 

I thought you were referring to the `testGetCourseIds` method, to get the course IDs for multiple lessons IDs. I got confused in the name of my own methods. :upside_down_face: 

I just pushed the commit 151b298005e69b65ccaa482a67044792d5211bfd fixing this. :)

renatho replied:
haha! no problem!

renatho commented on tests/unit-tests/test-class-lesson.php:
```diff
@@ -101,13 +100,110 @@ public function testIsPreRequisiteComplete() {
+				$course_id,
+				"Lesson with ID {$lesson_id} has course ID {$course_id}, expected {$expected_course_id}"
+			);
+			$this->assertEquals(
```
I think this test is the same as the previous one, just including this extra assertion, right? If that's correct, should we remove the previous test?

fjorgemota replied:
Yep, it's mostly for testing the cache. That's why it looks so similar.

I thought about making it the same, but I wasn't 100% sure. I'll change it then..

fjorgemota replied:
Wait, by "previous one" you mean `testGetCourseId`? If that's correct, it's two different tests: One verify against `get_course_id`, and another verifies against `get_course_ids` (note the `s` at the end)

renatho replied:
That's right! I added another comment above about the previous `testGetCourseId`.

fjorgemota replied:
Considering that `testGetCourseId` and `testGetCourseIds` are different tests for different methods, I don't feel like removing any of these tests. I removed the test that verified against the cache for the `testGetCourseIds` method, tho, in [d0963e5](https://github.com/Automattic/sensei/pull/5294/commits/d0963e517d031e03a26b97966f612fa48188cfa8).

renatho commented on tests/unit-tests/test-class-lesson.php:
```diff
@@ -101,13 +100,110 @@ public function testIsPreRequisiteComplete() {
+				"get_course_ids returned ID {$course_id} for lesson {$lesson_id}, but get_course_id returned {$get_course_id_result}"
+			);
+		}
+		shuffle( $lesson_ids );
```
And this test is the same as the previous one until this line, right? Should we extract this part to a private method here?

fjorgemota replied:
`testGetCourseIds` and `testGetCourseIdsCache` are basically the same, yeah.

I'm not sure about extracting this part to a private method, tho, because that means basically moving almost all of the `testGetCourseIds`, **including** the assertions, and just returning two variables: `$lesson_ids` and `$lesson_id_to_course_id`, as we need both of these variables to check against the cache.

 I think it could make sense to delete the `testGetCourseIds` and rename `testGetCourseIdsCache` to `testGetCourseIds`, but..I'm not sure. WDYT?

renatho replied:
> I think it could make sense to delete the testGetCourseIds and rename testGetCourseIdsCache to testGetCourseIds, but..I'm not sure. WDYT?

Might be! So we avoid the duplicated tests. 🙂

fjorgemota replied:
Done in [d0963e5](https://github.com/Automattic/sensei/pull/5294/commits/d0963e517d031e03a26b97966f612fa48188cfa8).

renatho commented on tests/unit-tests/test-class-lesson.php:
```diff
@@ -101,13 +100,110 @@ public function testIsPreRequisiteComplete() {
 
 	}
 
+	public function testGetCourseId() {
```
I think it would be nice to add some comments to every test. I think it's a practice we're using in Sensei codebase for new PHP tests.

fjorgemota replied:
Nice! Added some very basic documentation in [2014fc4](https://github.com/Automattic/sensei/pull/5294/commits/2014fc48a697c22bbfbec4019e8f2c88c87087ee). Hope it's...ok? :)

renatho replied:
That's perfect! I think we could just add the period after the descriptions. 😉

fjorgemota replied:
Done in [1f4fc6a](https://github.com/Automattic/sensei/pull/5294/commits/1f4fc6afcce4e8bce49166a39d35409433868ff7). Thanks!

renatho reviewed (commented):
Looking nice in general! I just added some comments for you. 😉

alexsanford commented on includes/class-sensei-lesson.php:
```diff
@@ -3822,6 +3822,53 @@ public function get_course_id( $lesson_id ) {
+	public function get_course_ids( $lesson_ids ) {
+		global $wpdb;
+
+		sort( $lesson_ids, SORT_NUMERIC );
```
Should we use `array_unique` here as well? 🤔

fjorgemota replied:
Good catch! Fixed in [bc9074b](https://github.com/Automattic/sensei/pull/5294/commits/bc9074bb7a65c9e161f14b9f1799ad02562c5083).

alexsanford replied:
Just to be slightly picky...I wonder if it's better to sort first, and then run `array_unique`? I'm not sure, but I believe `array_unique` has to do a sort anyway, so I'm guessing it might be quicker if the array is already sorted? 🤷‍♂️

fjorgemota replied:
Changed in [9ec32a1](https://github.com/Automattic/sensei/pull/5294/commits/9ec32a143b0faf39629852384eaa208ef94e3ddc).

I also added a `SORT_NUMERIC` parameter to `array_unique` to help with the sorting process. You're 100% right on your guess! :)

fjorgemota committed:
Apply md5 to lesson IDs on get course IDs

fjorgemota committed:
Make sure that the lesson IDs array has only unique items

fjorgemota committed:
Do not cache results on error in get_course_ids

fjorgemota committed:
Add some documentation for the new tests

fjorgemota committed:
Make the tests less random and more deterministic

fjorgemota committed:
Process sort before using array_unique

fjorgemota commented on includes/class-sensei-lesson.php:
```diff
@@ -3822,6 +3822,54 @@ public function get_course_id( $lesson_id ) {
+			foreach ( $results as $result ) {
+				$courses_by_lesson[ $result->lesson_id ] = $result->course_id;
+			}
+			wp_cache_set( $cache_key, $courses_by_lesson, $cache_group );
```
One thing that I just thought about: Right now, I'm calling `wp_cache_set` with the default expiration value, which is zero ([which means no expiration](https://developer.wordpress.org/reference/functions/wp_cache_set/)).

Should we define an expiration here? WDYT @renatho @aaronfc @alexsanford?

aaronfc replied:
humm... that's a good point.

Now I am wondering if we really need any cache here at all. In the end this is just a query (with multiple joins, which I have already given my opinion [here](https://github.com/Automattic/sensei/pull/5294#discussion_r913421183)) but there is no heavy processing anywhere.

My reasoning is:
this could change at any point by a user (assigning new lessons to a course or moving lessons between courses), so this **must not have unlimited expiration time**... 
I would use something low (minutes) to avoid problems with "old-fashioned" information.
An even better alternative would be intercepting the places where the underlying information can be changed and invalidate the cache actively there.
These are the usual issues with caches and I was wondering if we really need the cache at all.

fjorgemota replied:
hmmm...Yeah, I think the maximum expiration time here would be just a few minutes, maybe 1 or 5 minutes, something like that.

Also, I wouldn't personally leave this without ANY caching at all, just to avoid a possible database overload (even tho I doubt that's possible with such a simple query).

About cache invalidation...that would be waaay too annoying because we would have to guess which cache keys a given lesson is part of, which involves a big amount of tracking, and is just not worth doing at all in this case. I'll update it

alexsanford replied:
Yeah...really good point.

I think adding Lessons to Courses would be fine. The cache key is based on the array of Lesson IDs, so adding a new ID to the array shouldn't hit the old cached value.

I think the problem would be when an existing Lesson is moved to a new Course. Then it is associated with the old course in the cache, but nothing invalidates that cache.

What about caching in a variable of some sort instead? I'm thinking that if this method is called multiple times within a single request, it might be nice to return the result quickly.

But that could also potentially get complicated. Maybe it's best just not to have a cache? I'm not sure.

fjorgemota replied:
I think I just discovered a good way to fix that.

I can change the group used for caching here from `sensei` to something like `sensei/temporary` and set that group as non-persistent, using [wp_cache_add_non_persistent_groups](https://developer.wordpress.org/reference/functions/wp_cache_add_non_persistent_groups/).

That way the expiration will still be checked, but we have the warranty (well, sort of, the official implementation is empty because the native object cache is always temporary, so non-persistent) that the data will not be saved to some persistent cache or something, which means that it will only be saved for the duration of the request.

Does that sound good enough, cc @alexsanford @aaronfc? Sounds similar to caching in a variable of some sort, and I guess that would make phpcs happy anyway..

alexsanford replied:
Yeah, that sounds like a good solution to me 🤷

fjorgemota replied:
I added an expiration time here [bb0c389](https://github.com/Automattic/sensei/pull/5294/commits/bb0c389c961e40c5df22a6a867f1e188d0315680) and I set up a non-persistent cache group for Sensei here: [c50c233](https://github.com/Automattic/sensei/pull/5294/commits/c50c2339b57308fa6f6e69a0892eedc361bdd4f1)

Technically, the expiration time isn't needed anymore, but I guess it shoudln't be a problem to just leave it there.. :thinking:

fjorgemota committed:
Cache association for only 60 seconds (1 minute)

So we reduce the load on the database while also quickly expiring the association.

fjorgemota committed:
Add cache group "sensei/temporary" as a non-persistent cache group

renatho commented on includes/class-sensei-lesson.php:
```diff
@@ -3822,6 +3822,54 @@ public function get_course_id( $lesson_id ) {
+	 * @param array<int> $lesson_ids Array of Lesson IDs.
+	 * @return array<int, int|false> Array where the key is the lesson ID, and the value is the course ID (or false if not found)
+	 */
+	public function get_course_ids( $lesson_ids ) {
```
One think that I noticed is that it would be interesting to add a check in the `$lesson_ids`, like `empty( $lesson_ids )`, otherwise, if we call it with an empty array, we'll have a SQL syntax error in the `IN ()`

fjorgemota replied:
Excellent catch! Fixed in [eacc941](https://github.com/Automattic/sensei/pull/5294/commits/eacc941c8adfe06d61b854342498196510fcf58f).

fjorgemota committed:
Use wp_list_pluck to avoid foreach loop

fjorgemota committed:
Merge the array instead of replacing one with another

fjorgemota committed:
Add period on the end of each phpdoc comment

fjorgemota committed:
Delete duplicated method and make the test check against two things

fjorgemota committed:
Check against an empty list of $lesson_ids and return early to avoid exception

fjorgemota committed:
Remove $lesson_id_to_course_id in favour of recomputing the expected course ID

fjorgemota committed:
Format PHP code of the test-class-lesson.php file

fjorgemota committed:
Revert "Merge the array instead of replacing one with another"

This reverts commit 69bf0f1fdd3b6bdce2232fabe375e4afbdc292f0.

fjorgemota committed:
Revert "Use wp_list_pluck to avoid foreach loop"

This reverts commit 5918eabb6359ebee8a62d73a2c29cd6680e7d3a8.

fjorgemota committed:
Fix test to get course ID correctly and avoid issues with shuffled list

fjorgemota committed:
Avoid unnecessary loop in testGetCourseId

renatho reviewed (approved):
Looks good and works well to me! 🚀
As I can see in the discussions, I think all of them are done, right?
